### PR TITLE
Fix device parser logic to avoid creating CUDA context

### LINKED ISSUE #14292: Actor hangs with `ray` using PL v>=1.7
When invoking a training from within a Ray actor, the training process hangs. The reason for this is that Ray hangs (happens on multiple Ubuntu machines) when forking a process the way you did when calling `device_parser.num_cuda_devices`.  This doesn't happen on v1.6.5.

To sum-up (before delving into the details), the hanging occurs because of the following change between versions:
In v1.6.5 the `gpu` (changed to `cuda` in the newer version) accelerator just calls `torch.cuda.device_count() > 0` directly, while in 1.7.0 and above, this function `torch.cuda.device_count` is called in a forked process  (`utilities/device_parser.py`, line 346, in `num_cuda_devices`): 
```
with multiprocessing.get_context("fork").Pool(1) as pool:
    return pool.apply(torch.cuda.device_count)
```

The way I deduced the problem is by listening to system-signals and printing the traceback when getting `SIGUSR1` within the hanged process. The traceback we obtained was (starting after calling `pl.Trainer(...)`)

```
File "<__CONDA__ENV__>/lib/python3.9/site-packages/pytorch_lightning/utilities/argparse.py", line 345, in insert_env_defaults
  return fn(self, **kwargs)
File "<__CONDA__ENV__>/lib/python3.9/site-packages/pytorch_lightning/trainer/trainer.py", line 433, in __init__
  self._accelerator_connector = AcceleratorConnector(
File "<__CONDA__ENV__>/lib/python3.9/site-packages/pytorch_lightning/trainer/connectors/accelerator_connector.py", line 212, in __init__
  self._set_parallel_devices_and_init_accelerator()
File "<__CONDA__ENV__>/lib/python3.9/site-packages/pytorch_lightning/trainer/connectors/accelerator_connector.py", line 525, in _set_parallel_devices_and_init_accelerator
  if not self.accelerator.is_available():
File "<__CONDA__ENV__>/lib/python3.9/site-packages/pytorch_lightning/accelerators/cuda.py", line 91, in is_available
  return device_parser.num_cuda_devices() > 0
File "<__CONDA__ENV__>/lib/python3.9/site-packages/pytorch_lightning/utilities/device_parser.py", line 346, in num_cuda_devices
  return pool.apply(torch.cuda.device_count)
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/pool.py", line 736, in __exit__
  self.terminate()
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/pool.py", line 654, in terminate
  self._terminate()
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/util.py", line 224, in __call__
  res = self._callback(*self._args, **self._kwargs)
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/pool.py", line 729, in _terminate_pool
  p.join()
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/process.py", line 149, in join
  res = self._popen.wait(timeout)
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/popen_fork.py", line 43, in wait
  return self.poll(os.WNOHANG if timeout == 0.0 else 0)
File "<__CONDA__ENV__>/lib/python3.9/multiprocessing/popen_fork.py", line 27, in poll
  pid, sts = os.waitpid(self.pid, flag)
```

The hanging is reproduced by forking from within a Ray worker:
```python
import ray
import multiprocessing

@ray.remote
def f():
    with multiprocessing.get_context("fork").Pool(1) as pool:
        return pool.apply(int)

print(ray.get(f.remote()))
```

Thanks!

cc @tchaton @rohitgr7 @justusschock @kaushikb11 @awaelchli @akihironitta

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #14292 

Discussed several solutions offline with @carmocca and @justusschock:
- Rearrange logic in AcceleratorConnector: not feasible
- Use a different start method to launch the process in which we evaluate the cuda function: too costly
- Detect whether we are running inside of ray process: Not possible

Opened issue on torch side: https://github.com/pytorch/pytorch/issues/83973

The current solution seems to be the only one that allows us to have support for ddp-fork and at the same time be compatible with other frameworks. Ray users will have to set `PL_DISABLE_FORK=1` for now. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @tchaton @rohitgr7 @justusschock @kaushikb11 @awaelchli @akihironitta

### CODE REVIEW

File: src/pytorch_lightning/utilities/device_parser.py [hunk 1]
```diff
@@ -340,7 +341,7 @@ def num_cuda_devices() -> int:
     Unlike :func:`torch.cuda.device_count`, this function will do its best not to create a CUDA context for fork
     support, if the platform allows it.
     """
-    if "fork" not in torch.multiprocessing.get_all_start_methods():
+    if "fork" not in torch.multiprocessing.get_all_start_methods() or _is_forking_disabled():
```
Comment (justusschock) [inline src/pytorch_lightning/utilities/device_parser.py:344]: shouldn't this `if "fork" not in torch.multiprocessing.get_all_start_methods()` not be part of `_is_forking_disabled`? Doesn't matter if it's disabled due to the env variable or because it's not supported by the OS

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/utilities/device_parser.py (hunk 1 above)
Comment (awaelchli) [reply src/pytorch_lightning/utilities/device_parser.py:344]: Maybe it could, but we also have this kind of check in the accelerator connector 

```py
        if strategy_flag in _DDP_FORK_ALIASES and "fork" not in torch.multiprocessing.get_all_start_methods():
            raise ValueError(
                f"You selected `Trainer(strategy='{strategy_flag}')` but process forking is not supported on this"
                f" platform. We recommed `Trainer(strategy='ddp_spawn')` instead."
            )
```

for a specific user message about platform support. So in other words, at the moment "fork not supported by platform" and "fork disabled" are two different error messages.

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md [hunk 2]
```diff
@@ -18,6 +18,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added support for saving sharded optimizer state dict outside of `DDPShardedStrategy` ([#14208](https://github.com/PyTorchLightning/pytorch-lightning/pull/14208))
 
 
+- Added an environment variable `PL_DISABLE_FORK` that can be used to disable all forking in the Trainer ([#14319](https://github.com/Lightning-AI/lightning/issues/14319))
```
Comment (rohitgr7) [inline src/pytorch_lightning/CHANGELOG.md:21]: can you add it to the docs?

**Review by rohitgr7 — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md (hunk 2 above)
Comment (rohitgr7) [reply src/pytorch_lightning/CHANGELOG.md:21]: oh! nvm.
there's an error

Comment (awaelchli) [reply src/pytorch_lightning/CHANGELOG.md:21]: I don't want to. This is just a workaround until https://github.com/pytorch/pytorch/issues/83973 lands. And we will open an issue in ray lightning to set this variable.

Comment (justusschock) [reply src/pytorch_lightning/CHANGELOG.md:21]: @awaelchli I still would prefer it to be mentioned somewhere in the docs as even when the issue in torch lands, this would only be valid for the most recent torch version meaning that we may need to go with the current approach for a while (until all supported versions of torch are released past this issue).

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/CHANGELOG.md (hunk 2 above)
Comment (awaelchli) [reply src/pytorch_lightning/CHANGELOG.md:24]: Yes, we are aware that this will have to stay for a while. But every user is a potential ray user and so documenting this environment variable in the "introduction" is not exactly great. It also can't exactly be documented in the ddp-fork section because the purpose is different. Then the only other solution is to document ALL environment variables in Lightning in one place. This will unnecessarily hold back this PR.

Comment (justusschock) [reply src/pytorch_lightning/CHANGELOG.md:24]: Agreed. I think documenting ALL environment variables together with their purpose is something that would definitely be helpful but also definitely goes beyond this PR :)
